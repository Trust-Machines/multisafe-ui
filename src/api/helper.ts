import {ClarityAbi} from '@stacks/transactions';
import {NETWORK} from '@trustmachines/multisafe-contracts';

export const validateSafeAbi = (abi: ClarityAbi) => {
    const publicFns = [
        'add-owner',
        'remove-owner',
        'set-threshold',
        'submit',
        'confirm',
        'revoke'
    ]

    for (let p of publicFns) {
        if (!abi.functions.find(x => x.name === p && x.access === 'public')) {
            return false;
        }
    }

    return true;
}

export const makeTxUrl = (txId: string, network: NETWORK) => {
    return `https://explorer.stacks.co/txid/${txId}?chain=${network}`;
}