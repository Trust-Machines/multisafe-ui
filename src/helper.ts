import {ClarityAbi} from '@stacks/transactions';
import {NETWORK} from '@trustmachines/multisafe-contracts';
import BigNumber from 'bignumber.js';

export const validateSafeAbi = (abi: ClarityAbi) => {
    const publicFns = [
        'add-owner',
        'remove-owner',
        'set-min-confirmation',
        'submit',
        'confirm',
        'revoke'
    ]

    for (let p of publicFns) {
        if (!abi.functions.find(x => x.name === p && x.access === 'public')) {
            return false;
        }
    }

    return true;
}

export const makeTxUrl = (txId: string, network: NETWORK) => {
    return `https://explorer.stacks.co/txid/${txId}?chain=${network}`;
}

export const formatUnits = (value: BigNumber | string, decimals: number): BigNumber => {
    const bn = typeof value === 'string' ? new BigNumber(value) : value;
    return bn.dividedBy(10 ** decimals);
}

export const parseUnits = (value: string, decimals: number): BigNumber => {
    return new BigNumber(value).multipliedBy(10 ** decimals);
}