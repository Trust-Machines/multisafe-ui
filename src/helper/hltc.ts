import { script, Network, networks, payments } from 'bitcoinjs-lib';
import {
    bytesToHex,
    IntegerType,
    intToHexString,
    hexToBytes,
    utf8ToBytes,
} from 'micro-stacks/common';
import { hashSha256 } from 'micro-stacks/crypto-sha';

export interface HTLC {
    hash: Uint8Array;
    senderPublicKey: Uint8Array;
    recipientPublicKey: Uint8Array;
    expiration?: number;
    swapper: IntegerType;
}

export type BufferType = Buffer | Uint8Array;

export function numberToLE(num: IntegerType, length = 4) {
    const hexBE = intToHexString(num, length);
    let le = '';
    // reverse the buffer
    for (let i = 0; i < length; i++) {
        le += hexBE.slice(-2 * (i + 1), -2 * i || length * 2);
    }
    return le;
}

export function numberToLEBytes(num: IntegerType, length = 4) {
    return hexToBytes(numberToLE(num, length));
}

export const CSV_DELAY = 500;
export const CSV_DELAY_BUFF = script.number.encode(CSV_DELAY);
export const CSV_DELAY_HEX = CSV_DELAY_BUFF.toString('hex');

export function encodeExpiration(expiration?: number): Buffer {
    return typeof expiration === 'undefined' ? CSV_DELAY_BUFF : script.number.encode(expiration);
}

export function htlcASM({ hash, senderPublicKey, recipientPublicKey, expiration, swapper }: HTLC) {
    const swapperHex = numberToLE(swapper);
    return `
  ${swapperHex} OP_DROP
	OP_IF
    OP_SHA256 ${bytesToHex(hash)}
    OP_EQUALVERIFY
		${bytesToHex(recipientPublicKey)}
	OP_ELSE
		${encodeExpiration(expiration).toString('hex')}
		OP_CHECKSEQUENCEVERIFY
		OP_DROP
		${bytesToHex(senderPublicKey)}
	OP_ENDIF
	OP_CHECKSIG`
        .replace(/\s+/g, ' ')
        .trim();
}

export function generateHTLCScript(htlc: HTLC) {
    const asm = htlcASM(htlc);
    const output = script.fromASM(asm);
    return output;
}

export function generateHTLCAddress(htlc: HTLC, _network?: Network) {
    const network = _network || networks.regtest;
    const script = generateHTLCScript(htlc);
    // const payment = payments.p2wsh({ redeem: { output: script, network }, network });
    // const payment = payments.p2sh({ , network });
    // const hash = ripemd160(Uint8Array.from(script));
    // const payment = payments.p2sh({ output: Buffer.from(hash), network });
    const payment = payments.p2sh({ redeem: { output: script, network }, network });
    return payment;
}

export function reverseBuffer(buffer: BufferType): Uint8Array {
    if (buffer.length < 1) return buffer;
    let j = buffer.length - 1;
    let tmp = 0;
    for (let i = 0; i < buffer.length / 2; i++) {
        tmp = buffer[i];
        buffer[i] = buffer[j];
        buffer[j] = tmp;
        j--;
    }
    return Uint8Array.from(buffer);
}

export function getScriptHash(output: BufferType): Uint8Array {
    const uintOutput = Uint8Array.from(output);
    const hash = hashSha256(uintOutput);
    const reversed = reverseBuffer(Buffer.from(hash));
    return reversed;
}

export function hexPadded(hex: string) {
    const bytes = hex.length % 2 ? `0${hex}` : hex;
    return bytes;
}

export function secretToHash(secret: string) {
    const bytes = utf8ToBytes(secret);
    return hashSha256(bytes);
}
