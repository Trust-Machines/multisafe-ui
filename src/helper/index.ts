import BigNumber from 'bignumber.js';
import {DEPLOYERS, NETWORK} from '@trustmachines/multisafe-contracts';
import {StacksNetwork} from '@stacks/network';
import {contractPrincipalCV, validateStacksAddress} from '@stacks/transactions';
import {NETWORKS} from '../constants';
import {escapeRegExp} from '../util';
import {TransactionType} from '../types';

export const formatUnits = (value: BigNumber | string, decimals: number): BigNumber => {
    const bn = typeof value === 'string' ? new BigNumber(value) : value;
    return bn.dividedBy(10 ** decimals);
}

export const parseUnits = (value: string, decimals: number): BigNumber => {
    return new BigNumber(value).multipliedBy(10 ** decimals);
}

export const checkDecimalAmount = (input: string) => {
    const re = RegExp(`^\\d*(?:\\\\[.])?\\d*$`);

    if (!re.test(escapeRegExp(input))) {
        return false;
    }

    if (input.indexOf('.') !== -1) {
        return input.split('.')[1].length <= 6;
    }

    return true;
}

export const checkAmount = (input: string) => {
    return RegExp(/^(\s*|\d+)$/).test(input)
}

export const getStacksNetwork = (n: NETWORK): StacksNetwork => {
    return NETWORKS[n];
}

export const detectTransactionType = (executor: string): TransactionType => {
    const exec = executor.replace("'", '');
    for (const d of DEPLOYERS) {
        if (`${d}.add-owner` === exec) {
            return 'add-owner';
        }

        if (`${d}.remove-owner` === exec) {
            return 'remove-owner';
        }

        if (`${d}.set-threshold` === exec) {
            return 'set-threshold';
        }

        if (`${d}.transfer-stx` === exec) {
            return 'transfer-stx';
        }

        if (`${d}.transfer-sip-009` === exec) {
            return 'transfer-sip-009';
        }

        if (`${d}.transfer-sip-010` === exec) {
            return 'transfer-sip-010';
        }
    }

    return null;
}

export const contractPrincipalCVFromString = (s: string) => {
    const [a, b] = s.split(".");
    return contractPrincipalCV(a, b);
}

export const transformNftUri = (uri: string, nftId: string) => {
    const u = uri.replace('{id}', nftId);

    if (u.startsWith('ipfs://')) {
        return u.replace(/ipfs:\/\/(ipfs\/)?/, 'https://gateway.ipfs.io/ipfs/');
    }

    if (u.startsWith('ar://')) {
        return u.replace('ar://', 'https://arweave.net/');
    }

    return u;
}

export const isValidRecipient = (r: string) => {
    if (r === '') {
        return false;
    }

    if (validateStacksAddress(r)) {
        return true;
    }

    try {
        contractPrincipalCVFromString(r);
        return true;
    } catch (e) {

    }

    return false;
}
