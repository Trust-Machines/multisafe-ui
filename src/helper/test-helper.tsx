import {render} from '@testing-library/react';
import {createHistory, createMemorySource, LocationProvider} from '@reach/router';

export const renderWithRouter = (
    ui: any,
    {route = '/', history = createHistory(createMemorySource(route))} = {},
) => {
    return {
        ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
        // adding `history` to the returned utilities to allow us
        // to reference it in our tests (just try to avoid using
        // this to test implementation details).
        history,
    }
}