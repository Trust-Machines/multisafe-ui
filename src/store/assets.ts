import {atom} from 'jotai';

import {FTAsset, NFTAsset} from '../types';

export interface FTListItem extends FTAsset {
    type: "ft"
}

export interface NFTListItem extends NFTAsset {
    type: "nft"
}

export type ListItem = FTListItem | NFTListItem;

export interface AssetsState {
    loading: boolean,
    list: ListItem[],
    init: boolean
}

export const assetsAtom = atom<AssetsState>({loading: false, list: [], init: false});