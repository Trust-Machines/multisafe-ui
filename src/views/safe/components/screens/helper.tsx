import React from 'react';
import Typography from '@mui/material/Typography';
import {Box, useTheme} from '@mui/material';

const ScreenHeader = (props: { title: string, icon: JSX.Element, children: JSX.Element }) => {
    const theme = useTheme();
    const icon = React.cloneElement(props.icon, {
        sx: {
            marginRight: '4px',
            color: theme.palette.primary.main
        }
    })
    return <Box sx={{
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'space-between',
        mb: '20px'
    }}>
        <Typography variant="h6" sx={{display: 'flex', alignItems: 'center',}}>
            {icon} {props.title}
        </Typography>
        {props.children}
    </Box>
}

export default ScreenHeader;