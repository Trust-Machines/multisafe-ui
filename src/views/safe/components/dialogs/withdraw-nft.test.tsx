import WithdrawNft from './withdraw-nft';
import {renderWithRouter} from '../../../../helper/test-helper';
import assets from '../../../../constants/nft-list';

test('1 Default Render', () => {
    const view = renderWithRouter(<WithdrawNft asset={assets['mainnet'][1]}/>);
    expect(view.container).toMatchSnapshot();
})

test('2 Render with nftId', () => {
    const view = renderWithRouter(<WithdrawNft asset={assets['mainnet'][1]} nftId="23"/>);
    expect(view.container).toMatchSnapshot();
})