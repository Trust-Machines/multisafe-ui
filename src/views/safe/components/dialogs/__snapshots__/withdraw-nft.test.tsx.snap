// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 Default Render 1`] = `
<div>
  <h2
    class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
  >
    Withdraw Bitcoin Monkeys
    <button
      aria-label="close"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-lc2yqr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </h2>
  <div
    class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
  >
    <div
      class="MuiBox-root css-542elh"
    >
      <div
        class="MuiBox-root css-uewl2b"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-1"
            id="mui-1-label"
          >
            NFT ID
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby="mui-1-helper-text"
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-1"
              maxlength="30"
              name="nft-id"
              placeholder="Enter ID of the NFT you like to transfer"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1in441m"
              >
                <span>
                  NFT ID
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-focused css-1wc848c-MuiFormHelperText-root"
            id="mui-1-helper-text"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </p>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-2"
          id="mui-2-label"
        >
          Recipient
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-describedby="mui-2-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-2"
            name="recipient"
            placeholder="Recipient"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1ftyaf0"
            >
              <span>
                Recipient
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
          id="mui-2-helper-text"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </p>
      </div>
    </div>
  </div>
  <div
    class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Cancel
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Withdraw
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`2 Render with nftId 1`] = `
<div>
  <h2
    class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
  >
    Withdraw Bitcoin Monkeys
    <button
      aria-label="close"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-lc2yqr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </h2>
  <div
    class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
  >
    <div
      class="MuiBox-root css-542elh"
    >
      <div
        class="MuiBox-root css-uewl2b"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-3"
            id="mui-3-label"
          >
            NFT ID
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-describedby="mui-3-helper-text"
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-3"
              maxlength="30"
              name="nft-id"
              placeholder="Enter ID of the NFT you like to transfer"
              type="text"
              value="23"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1in441m"
              >
                <span>
                  NFT ID
                </span>
              </legend>
            </fieldset>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained Mui-focused MuiFormHelperText-filled css-1wc848c-MuiFormHelperText-root"
            id="mui-3-helper-text"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </p>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-4"
          id="mui-4-label"
        >
          Recipient
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-describedby="mui-4-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-4"
            name="recipient"
            placeholder="Recipient"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1ftyaf0"
            >
              <span>
                Recipient
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
          id="mui-4-helper-text"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </p>
      </div>
    </div>
  </div>
  <div
    class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Cancel
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Withdraw
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
